### PR TITLE
6746 network data

### LINKED ISSUE #6746: Reimagine `message NodeData{}`
# Problem encountered
The data-node returns the following network information from its `rpc GetNetworkData()` endpoint:

```
message NodeData {
  // Total staked amount across all nodes
  string taked_total = 1;
  // Total number of nodes
  uint32 total_nodes = 2;
  // Number of inactive nodes
  uint32 inactive_nodes = 3;
  // Number of nodes validating
  uint32 validating_nodes = 4;
  // Total uptime for all epochs across all nodes
  float uptime = 5;
}
```

This structure existed long before the validators-joining-and-leaving work and doesn't necessarily "fit" against what nodes do anymore.

For example what is an "inactive" node? Currently `inactive_nodes` returns the count of all pending/ersatz nodes which doesn't seem right -- you can be an up and running, fully participating erstaz node following all the rules but counted as inactive.

We should think about what information people would expect to see as an overview of the network and change `NodeData` to match this.

# Observed behaviour
`NodeData` returns a field for `inactive_nodes` with no description of what this means.

# Expected behaviour
`NodeData` returns useful information about the network with clear doc-strings for what each field represents.

# System response
Describe what the system response was, include the output from the command, automation, or else.

# Evidence

## Logs
If applicable, add logs and/or screenshots to help explain your problem.

## Additional context
Add any other context about the problem here including; system version numbers, components affected.

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All bug recreation steps can be followed without presenting the original error/bug
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new [system-tests](https://github.com/vegaprotocol/system-tests/issues)
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

### PR DESCRIPTION
**Pull request by wwestgarth**
closes #6746 

This changes what we return in `NodeData` to better reflect how the network topology looks with all the new tendermint/ersatz/pending states a node can have. The field `uint32 validating_nodes = 4` has been replaced with:

```
NodeSet tendermint_nodes = 4;
NodeSet ersatz_nodes = 5;
NodeSet pending_nodes = 6;
```

where for each `NodeSet` you get the folloing details:
```
message NodeSet {
  // Total number of nodes in the nodeset
  uint32 total = 1;
  // Number of nodes in the nodeset which had a performance score of 0 at the end of the last epoch
  uint32 inactive = 2;
  // IDs of nodes that were promoted into this nodeset at the start of the epoch
  repeated string promoted = 3;
  // IDs of nodes that were demoted into this nodeset at the start of the epoch
  repeated string demoted = 4;
  // Total number of nodes allowed in the nodeset
  optional uint32 maximum = 5;
}
```

The idea is that this information, at a glance, will tell you things about how the nodes are moving around each epoch. For example if for a nodeset `maximum > total` you know there are empty slots and someone could join, if a netparam changed and so for a nodeset `maximum < total` you know someone is going to be pushed out a demoted. You can then see for an epoch who got demoted/promoted into which set.

I've wired it into some system-tests here: 
https://github.com/vegaprotocol/system-tests/compare/get-network-data

and heres what it looks like in GQL after starting with 5 TM nodes and changing some parameter so 2 get demoted:
Screenshot 2022-11-17 at 09 42 38

full system test run: https://jenkins.ops.vega.xyz/blue/organizations/jenkins/common%2Fsystem-tests/detail/system-tests/13342/pipeline

### CODE REVIEW

**Comment by jeremyletang** [comment]
This looks great. let's get it in before the release as well if possible!

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -856,19 +856,42 @@ type Epoch {
   ): DelegationsConnection
 }
 
+type NodeSet {
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:859]: Could you add in a line explaining what a node set is here? Also, I'm suggesting node set as 2 words below, but if you vehemently object, you can reject.

File: datanode/gateway/graphql/schema.graphql [hunk 2]
```diff
@@ -856,19 +856,42 @@ type Epoch {
...
+  "Total number of nodes in the nodeset"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:860]:
```suggestion
  "Total number of nodes in the node set"
```

File: datanode/sqlstore/node.go [hunk 3]
```diff
@@ -215,61 +214,61 @@ func (store *Node) GetNodeData(ctx context.Context) (entities.NodeData, error) {
...
-					ROW_NUMBER() OVER () AS id
-				FROM
-					nodes
-        LEFT JOIN ranking_scores ON ranking_scores.node_id = nodes.id AND
-                  ranking_scores.epoch_seq = (SELECT max(epoch_seq) FROM ranking_scores WHERE node_id = nodes.id)
-        WHERE EXISTS (
-					SELECT *
-					FROM join_event WHERE node_id = nodes.id
-				)
+					epoch_id = $1
```
Comment (jeremyletang) [inline datanode/sqlstore/node.go:217]: wow, isn't that a much nicer query here?

File: datanode/entities/enums.go [hunk 4]
```diff
@@ -723,6 +723,14 @@ const (
 	ValidatorNodeStatusPending     = ValidatorNodeStatus(vega.ValidatorNodeStatus_VALIDATOR_NODE_STATUS_PENDING)
 )
 
+// ValidatorStatusRanking so we know which direction was a promotion and which was a demotion.
+var ValidatorStatusRanking = map[ValidatorNodeStatus]int{
```
[the remark this answers is not in the document]
Comment (wwestgarth) [reply datanode/entities/enums.go:727]: Yea agree it is a bit over-engineering, I was more thinking about if in the future we add a new status type, or rename one and the proto enums then become out of order.

File: protos/sources/vega/vega.proto [hunk 5]
```diff
@@ -1226,17 +1226,35 @@ message Node {
...
+  // Total number of nodes allowed in the nodeset
+  optional uint32 maximum = 5;
+
+}
+
 message NodeData {
   // Total staked amount across all nodes
   string staked_total = 1;
-  // Total number of nodes
+  // Total number of nodes across all nodesets
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1246]:
```suggestion
  // Total number of nodes across all node sets
```

File: protos/sources/vega/vega.proto [hunk 6]
```diff
@@ -1226,17 +1226,35 @@ message Node {
...
   uint32 total_nodes = 2;
-  // Number of inactive nodes
+  // Total number of nodes which had a performance score of 0 at the end of the last epoch
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1248]:
```suggestion
  // Total number of nodes that had a performance score of 0 at the end of the last epoch
```

File: protos/sources/vega/vega.proto [hunk 7]
```diff
@@ -1226,17 +1226,35 @@ message Node {
...
   uint32 inactive_nodes = 3;
-  // Number of nodes validating
-  uint32 validating_nodes = 4;
+  // Details on the set of Tendermint nodes in the network
+  NodeSet tendermint_nodes = 4;
+  // Details on the set of Erstaz nodes in the network
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1252]:
```suggestion
  // Details on the set of ersatz (standby) nodes in the network
```

File: datanode/api/trading_data_v2.go [hunk 8]
```diff
@@ -2196,13 +2196,45 @@ func (t *tradingDataServiceV2) marketExistsForID(ctx context.Context, marketID s
...
+	if err != nil {
+		return nil, apiError(codes.Internal, err)
+	}
+
+	np, err = t.networkParameterService.GetByKey(ctx, "network.validators.ersatz.multipleOfTendermintValidators")
 	if err != nil {
 		return nil, apiError(codes.Internal, err)
 	}
 
+	erstazFactor, err := strconv.ParseFloat(np.Value, 32)
```
Comment (candida-d) [inline datanode/api/trading_data_v2.go:2227]: should be ersatz (not erstaz)

File: datanode/entities/enums.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyletang) [reply datanode/entities/enums.go:727]: fair enough

**Review by dexturr — DISMISSED** [review]

File: datanode/gateway/graphql/schema.graphql (hunk 1 above)
Comment (wwestgarth) [reply datanode/gateway/graphql/schema.graphql:860]: I have no preference! Updated to "node set".

File: datanode/gateway/graphql/schema.graphql [hunk 9]
```diff
@@ -856,19 +856,43 @@ type Epoch {
...
-  "Number of inactive nodes"
+  "Total number of nodes that had a performance score of 0 at the end of the last epoch"
   inactiveNodes: Int!
 
-  "Number of nodes validating"
-  validatingNodes: Int!
+  "Details on the set of consensus nodes in the network"
+  tendermintNodes: NodeSet!
+  
+  "Details on the set of erstaz (standby) nodes in the network"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:891]:
```suggestion
  "Details on the set of ersatz (standby) nodes in the network"
```
sorry  - I missed one

**Review by candida-d — DISMISSED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**